### DISCUSSION TITLE
Leafmap foliumap not plotting raster Sagemaker Studio

### DISCUSSION THREAD

**Discussion by mthsdiniz-usp** [opengeos/leafmap#714] [Q&A]
I'm working of Sagemaker Studio and trying to plot a few rasters using leafmap. 

As per the tutorials available on the documentation, i'm importing leafmap with the foliumap backend and trying to use the add_raster method to display on the map. When the code is executed, the map loads, the layer name shows on the layer pane but the image is not being displayed. 

Any ideas of why this might be happening? 

```
import leafmap.foliumap as leafmap

m = leafmap.Map()

m.add_tile_layer(
    url="https://mt1.google.com/vt/lyrs=y&x={x}&y={y}&z={z}",
    name="Google Satellite",
    attribution="Google",
)

m.add_raster('../rasters/teste.tif', bands=[1], layer_name="DEM")

m
```

Ps: I zipped the raster i'm using because github doesn't let me attach it here
[teste (4).zip](https://github.com/opengeos/leafmap/files/14960774/teste.4.zip)

**Comment by giswqs** [comment]
Did you install jupyter-server-proxy on SageMaker Studio? You can also check the chrome netowrk tab to see what the error is.

**Reply by mthsdiniz-usp** [reply]
I've installed jupyter-server-proxy using pip. 

The network tab is just giving me a connection_refused message:

GET http://127.0.0.1:37309/api/tiles/11/677/1213.png?&filename=%2Fhome%2Fsagemaker-user%2Fbags-model%2Fnotebooks%2F2024-01-26_aot.tif net::ERR_CONNECTION_REFUSED

**Reply by mthsdiniz-usp** [reply]
I think this might be related to the proxy configuration for the Sagemaker Studio (**not the Studio Lab**). Any ideas how to set the 'LOCALTILESERVER_CLIENT_PREFIX' for the Sagemaker Studio?

**Reply by giswqs** [reply]
Try adding the following before using add_raster. 

```python
import os
os.environ['LOCALTILESERVER_CLIENT_PREFIX'] = 'proxy/{port}'
```

**Comment by giswqs** [comment]
Actually, it should be:

```python
import os
os.environ['LOCALTILESERVER_CLIENT_PREFIX'] = f"studiolab/default/jupyter/proxy/{{port}}"
```

See [this](https://github.com/banesullivan/localtileserver/issues/128#issuecomment-1419770907).

**Reply by giswqs** [reply]
Another relevant issue [here](https://github.com/banesullivan/localtileserver/issues/101#issuecomment-1650817418).

**Comment by mthsdiniz-usp** [comment]
> Actually, it should be:
> 
> ```python
> import os
> os.environ['LOCALTILESERVER_CLIENT_PREFIX'] = f"studiolab/default/jupyter/proxy/{{port}}"
> ```
> 
> See [this](https://github.com/banesullivan/localtileserver/issues/128#issuecomment-1419770907).

With that string or the one before, i end up getting the same 404 error. I saw the issue you posted on the localtileserver but was wondering if you ended up finding the string to use with **Sagemaker Studio (Paid)** instead of the Studio Lab.

`GET https://d-a7fjd1r8yf0g.studio.us-east-1.sagemaker.aws/studiolab/default/jupyter/proxy/46651/api/tiles/11/678/1214.png?&filename=%2Fhome%2Fsagemaker-user%2Fbags-model%2Fnotebooks%2F2024-01-26_aot.tif 404 (Not Found)`

**Reply by giswqs** [reply]
This is tricky issue. I have not figured it out. @banesullivan might knows more about this.

**Reply by mthsdiniz-usp** [reply]
with the following line i don't get any erros on the console, but the raster doesn't load as well.

`os.environ['LOCALTILESERVER_CLIENT_PREFIX'] = f"jupyter/default/lab/proxy/{{port}}"`

My URL is https://d-a7fjd1r8yf0g.studio.us-east-1.sagemaker.aws/jupyter/default/lab/tree/bags-model/notebooks/3-ground_truth.ipynb

**Reply by banesullivan** [reply]
hmm, generally this looks right. Though I've never used sagemeaker so I may need to go on there and give this a try myself

@giswqs, have you gotten this to work on Sagemaker before?

**Reply by giswqs** [reply]
I have only managed to make it work with the free SageMaker Studio Lab, not the paid version of SageMaker Studio.

**Reply by banesullivan** [reply]
> with the following line i don't get any erros on the console, but the raster doesn't load as well.
> 
> `os.environ['LOCALTILESERVER_CLIENT_PREFIX'] = f"jupyter/default/lab/proxy/{{port}}"`
> 
> My URL is https://d-a7fjd1r8yf0g.studio.us-east-1.sagemaker.aws/jupyter/default/lab/tree/bags-model/notebooks/3-ground_truth.ipynb

Aha, I think I see the issue here. Can you try:

```py
os.environ['LOCALTILESERVER_CLIENT_PREFIX'] = f"jupyter/default/proxy/{{port}}"
```

`proxy` is another service like `lab` is for the actual `JupyterLab` service

**Reply by mthsdiniz-usp** [reply]
Hey @banesullivan , just tried setting things like that and got the following:

When trying the tileserver example, i got **500 (Internal Server Error)** on the console:

`GET https://d-a7fjd1r8yf0g.studio.us-east-1.sagemaker.aws/jupyter/default/proxy/39135/api/tiles/17/42582/58862.png?&filename=%2Fvsicurl%3Furl%3Dhttps%253A%252F%252Foin-hotosm.s3.amazonaws.com%252F59c66c5223c8440011d7b1e4%252F0%252F7ad397c0-bba2-4f98-a08a-931ec3a6e943.tif%26use_head%3Dno%26list_dir%3Dno 500 (Internal Server Error)`

```
from localtileserver import get_folium_tile_layer, TileClient, examples
from folium import Map
import os
os.environ['LOCALTILESERVER_CLIENT_PREFIX'] = f"jupyter/default/proxy/{{port}}"
# First, create a tile server from local raster file
# client = TileClient('path/to/geo.tif')
client = examples.get_oam2()  # use example data

# Create folium tile layer from that server
t = get_folium_tile_layer(client)

m = Map(location=client.center(), zoom_start=16)
m.add_child(t)
m
```

When trying with Leafmap, i ended up getting the same error:

`GET https://d-a7fjd1r8yf0g.studio.us-east-1.sagemaker.aws/jupyter/default/proxy/39135/api/tiles/11/676/1212.png?&filename=%2Froot%2Fbags-model%2Fnotebooks%2F2024-01-26_aot.tif&colormap=terrain 500 (Internal Server Error)`

```
import leafmap.foliumap as leafmap
import os
os.environ['LOCALTILESERVER_CLIENT_PREFIX'] = f"jupyter/default/proxy/{{port}}"
m = leafmap.Map()
m.add_raster('2024-01-26_aot.tif', palette="terrain",layer_name="DEM")
m
```

Below is the localtileserver report and i also attached the tif again.
[2024-01-26_aot.zip](https://github.com/opengeos/leafmap/files/15024367/2024-01-26_aot.zip)


```
--------------------------------------------------------------------------------
  Date: Fri Apr 12 15:31:21 2024 UTC

                OS : Linux
            CPU(s) : 32
           Machine : x86_64
      Architecture : 64bit
               RAM : 249.1 GiB
       Environment : Jupyter

  Python 3.11.1 (main, Feb  4 2023, 11:13:32) [GCC 10.2.1 20210110]

   localtileserver : 0.10.2
             click : 8.1.7
             flask : 3.0.3
     flask_caching : 2.1.0
        flask_cors : 4.0.0
       flask_restx : 1.3.0
             numpy : 1.26.1
          rasterio : 1.3.8.post2
         rio_cogeo : 5.3.0
         rio_tiler : 6.4.6
     server_thread : 0.2.0
          werkzeug : 3.0.2
            folium : 0.16.0
        ipyleaflet : 0.18.2
        jupyterlab : 4.1.6
        matplotlib : 3.8.0
           shapely : 2.0.2
         traitlets : 5.11.2
--------------------------------------------------------------------------------
```

@giswqs Does plotting cog's also leverage the localtileserver backend?

I tried executing the cog example and it worked just fine with and without setting the environment variable

![image](https://github.com/opengeos/leafmap/assets/86017219/bf464d5b-bca4-409c-9801-f5f1e72359eb)

**Reply by giswqs** [reply]
The `Map.add_cog_layer()` method uses [TiTiler](https://developmentseed.org/titiler/) under the hood, which uses the https://titiler.xyz endpoint. It works because it does not require running a server where the Jupyter instance is hosted. The downside is that the datasets need to be accessible through http. The `Map.add_raster()` method relies on localtileserver.

**Reply by mthsdiniz-usp** [reply]
@giswqs Makes sense. If we can't figure out why i'm getting those errors, i think i'm gonna add a s3 bucket upload step and temp http link generation to use Leafmap. My idea here is to use leafmap along with a basemap to introduce a human-in-the-loop step into Machine Learning algorithms. The ideia is that the user can simple use the pixel inspector to get pixel values for both positive samples and negative samples of a binary classification model.
